README.md
# Calypso for BGA

Calypso is available to play on Board Game Arena (as a beta version):

[<img src="./misc/Bga_button.png">](https://boardgamearena.com/gamepanel?game=calypso)

[Credits](./misc/credits.md)

## Dev

For devvy stuff check out [dev notes](./misc/dev.md).

## To-do list

Archived items in [an archive](./misc/archive.md), and [possible things to add](./misc/future_possibilities.md).

## Browser compatibility

Checked:

* Firefox (85.0.1)
* Edge (88.0.705.63)
* Chrome (88.0.4324.150)


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) (mostly).

## [Unreleased]

## [beta-v02] - 2023-07-30

### Added

- User preference to highlight playable cards (vs non-playable). Default is on

### Fixed

- Fixed spectator (silent) server errors by skipping problematic functions

## [beta-v01] - 2023-07-15

Entered beta on 15/7 - no changes

## [alpha-v06] - 2023-06-14

### Fixed

- Sorted introduced translation error

## [alpha-v05] - 2023-06-14

### Changed

- Nicer post-trick animations [PR #25](https://github.com/ADBond/bga-calypso/pull/25)

## [alpha-v04] - 2023-05-30

### Added

- Links to player-profiles from Calypso-mat names [PR #20](https://github.com/ADBond/bga-calypso/pull/20)
- Renounce-flag tooltips specialised to player name, trumps, and suit renounced [PR #21](https://github.com/ADBond/bga-calypso/pull/21)
- User preference option for four-colour card pack [PR #23](https://github.com/ADBond/bga-calypso/pull/23)
- Added changelog to keep track of changes more easily

### Changed

- Better card image quality [PR #22](https://github.com/ADBond/bga-calypso/pull/22)

## [alpha-v03] - 2023-05-27

### Added

- Sort hand relative to personal trumps [PR #18](https://github.com/ADBond/bga-calypso/pull/18)

## [alpha-v02] - 2023-05-25

### Added

 - User preference for table colour [PR #16](https://github.com/ADBond/bga-calypso/pull/16)

## [alpha-v01] - 2021-03-09

Online version working 2021-02-24.

### Added

- Working version of Calypso. Sorry this is as detailed as it will go up this point! Have a shufti at previous PRs and commit history if you really need more.

[unreleased]: https://github.com/ADBond/bga-calypso/compare/beta-v02...HEAD
[beta-v02]: https://github.com/ADBond/bga-calypso/compare/beta-v01...beta-v02
[beta-v01]: https://github.com/ADBond/bga-calypso/compare/alpha-v06...beta-v01
[alpha-v06]: https://github.com/ADBond/bga-calypso/compare/alpha-v05...alpha-v06
[alpha-v05]: https://github.com/ADBond/bga-calypso/compare/alpha-v04...alpha-v05
[alpha-v04]: https://github.com/ADBond/bga-calypso/compare/alpha-v03...alpha-v04
[alpha-v03]: https://github.com/ADBond/bga-calypso/compare/alpha-v02...alpha-v03
[alpha-v02]: https://github.com/ADBond/bga-calypso/compare/alpha-v01...alpha-v02
[alpha-v01]: https://github.com/ADBond/bga-calypso/releases/tag/alpha-v01


stats.inc.php
<?php

/**
 *------
 * BGA framework: © Gregory Isabelli <gisabelli@boardgamearena.com> & Emmanuel Colin <ecolin@boardgamearena.com>
 * Calypso implementation : © Andy Bond <48208438+ADBond@users.noreply.github.com>
 *
 * This code has been produced on the BGA studio platform for use on http://boardgamearena.com.
 * See http://en.boardgamearena.com/#!doc/Studio for more information.
 * -----
 *
 * stats.inc.php
 *
 * Calypso game statistics description
 *
 */

/*
    In this file, you are describing game statistics, that will be displayed at the end of the
    game.
    
    !! After modifying this file, you must use "Reload  statistics configuration" in BGA Studio backoffice
    ("Control Panel" / "Manage Game" / "Your Game")
    
    There are 2 types of statistics:
    _ table statistics, that are not associated to a specific player (ie: 1 value for each game).
    _ player statistics, that are associated to each players (ie: 1 value for each player in the game).

    Statistics types can be "int" for integer, "float" for floating point values, and "bool" for boolean
    
    Once you defined your statistics there, you can start using "initStat", "setStat" and "incStat" method
    in your game logic, using statistics names defined below.
    
    !! It is not a good idea to modify this file when a game is running !!

    If your game is already public on BGA, please read the following before any change:
    http://en.doc.boardgamearena.com/Post-release_phase#Changes_that_breaks_the_games_in_progress
    
    Notes:
    * Statistic index is the reference used in setStat/incStat/initStat PHP method
    * Statistic index must contains alphanumerical characters and no space. Example: 'turn_played'
    * Statistics IDs must be >=10
    * Two table statistics can't share the same ID, two player statistics can't share the same ID
    * A table statistic can have the same ID than a player statistics
    * Statistics ID is the reference used by BGA website. If you change the ID, you lost all historical statistic data. Do NOT re-use an ID of a deleted statistic
    * Statistic name is the English description of the statistic as shown to players
    
*/

$stats_type = array(

    // Statistics global to table
    "table" => array(

        "average_calypsos_per_round" => array(
            "id" => 10,
            "name" => totranslate("Total calypsos per round"),
            "type" => "float",
        ),
        "average_points_per_round" => array(
            "id" => 11,
            "name" => totranslate("Average individual score per round"),
            "type" => "float",
        ),

        // may as well keep analagous ids as for player stats
        "fastest_calypso" => array(
            "id" => 35,
            "name" => totranslate("Fastest calypso (by number of tricks)"),
            "type" => "int"
        ),

        "proportion_tricks_won_trump_lead" => array(
            "id" => 41,
            "name" => totranslate("Proportion of tricks won by leading trumps"),
            "type" => "float",
        ),
        "proportion_tricks_won_first_trump" => array(
            "id" => 42,
            "name" => totranslate("Proportion of tricks won by trumping in first"),
            "type" => "float",
        ),
        "proportion_tricks_won_overtrump" => array(
            "id" => 43,
            "name" => totranslate("Proportion of tricks won by overtrumping"),
            "type" => "float",
        ),
        "proportion_tricks_won_plainsuit" => array(
            "id" => 44,
            "name" => totranslate("Proportion of tricks won by highest plainsuit card"),
            "type" => "float",
        ),

    ),
    
    // Statistics existing for each player
    "player" => array(
        // per round statistics
        "calypsos_per_round" => array(
            "id" => 10,
            "name" => totranslate("Average calypsos per round (individual)"),
            "type" => "float"
        ),
        "partnership_calypsos_per_round" => array(
            "id" => 11,
            "name" => totranslate("Average calypsos per round (partnership)"),
            "type" => "float"
        ),
        "calypso_points_per_round" => array(
            "id" => 12,
            "name" => totranslate("Average calypso points per round (individual)"),
            "type" => "float",
        ),
        "partnership_calypso_points_per_round" => array(
            "id" => 13,
            "name" => totranslate("Average calypso points per round (partnership)"),
            "type" => "float",
        ),
        "incomplete_calypso_cards_per_round" => array(
            "id" => 14,
            "name" => totranslate("Average incomplete calypso cards per round (individual)"),
            "type" => "float",
        ),
        "partnership_incomplete_calypso_cards_per_round" => array(
            "id" => 15,
            "name" => totranslate("Average incomplete calypso cards per round (partnership)"),
            "type" => "float",
        ),
        "trickpile_cards_per_round" => array(
            "id" => 16,
            "name" => totranslate("Average trickpile cards per round (individual)"),
            "type" => "float",
        ),
        "partnership_trickpile_cards_per_round" => array(
            "id" => 17,
            "name" => totranslate("Average trickpile cards per round (partnership)"),
            "type" => "float",
        ),
        "points_per_round" => array(
            "id" => 18,
            "name" => totranslate("Average points per round (individual)"),
            "type" => "float",
        ),
        "partnership_points_per_round" => array(
            "id" => 19,
            "name" => totranslate("Average points per round (partnership)"),
            "type" => "float",
        ),
        
        // "total_cards_won" => array(
        //     "id" => 20,
        //     "name" => totranslate("Total captured cards per round (individual)"),
        //     "type" => "float",
        // ),

        "personal_trumps_per_hand" => array(
            "id" => 30,
            "name" => totranslate("Average personal trumps dealt per hand"),
            "type" => "float",
        ),
        "partner_trumps_per_hand" => array(
            "id" => 31,
            "name" => totranslate("Average partner trumps dealt per hand"),
            "type" => "float",
        ),
        "opponent_trumps_per_hand" => array(
            "id" => 32,
            "name" => totranslate("Average opponent trumps dealt per hand"),
            "type" => "float",
        ),

        "fastest_calypso" => array(
            "id" => 35,
            "name" => totranslate("Fastest calypso (by number of tricks)"),
            "type" => "int"
        ),

        "tricks_won_total_per_hand" => array(
            "id" => 40,
            "name" => totranslate("Average tricks won per hand"),
            "type" => "float",
        ),
        "tricks_won_trump_lead_per_hand" => array(
            "id" => 41,
            "name" => totranslate("Average tricks won by leading trumps per hand"),
            "type" => "float",
        ),
        "tricks_won_first_trump_per_hand" => array(
            "id" => 42,
            "name" => totranslate("Average tricks won by trumping in first per hand"),
            "type" => "float",
        ),
        "tricks_won_overtrump_per_hand" => array(
            "id" => 43,
            "name" => totranslate("Average tricks won by overtrumping per hand"),
            "type" => "float",
        ),
    
        "score_first_leader" => array(
            "id" => 50,
            "name" => totranslate("Individual score when first leader in first hand of round"),
            "type" => "int"
        ),
        "score_player_two" => array(
            "id" => 51,
            "name" => totranslate("Individual score when second player in first hand of round"),
            "type" => "int"
        ),
        "score_player_three" => array(
            "id" => 52,
            "name" => totranslate("Individual score when third player in first hand of round"),
            "type" => "int"
        ),
        "score_dealer" => array(
            "id" => 53,
            "name" => totranslate("Individual score when dealer in first hand of round"),
            "type" => "int"
        ),
    )
);
